File: frontend/src/routes/Inventory/lib/stockCalc.ts
import { ITEMS, getItem } from '../constants';

const RED_THRESHOLD = 8;    // < 2 weeks of sessions
const AMBER_THRESHOLD = 16; // < 4 weeks of sessions

export function sessionsRemaining(code: string, qty: number): number | null {
  if (code === 'PAK-001') return qty * 10;
  if (code === 'P00012326') return Math.floor(qty / 2);
  const item = getItem(code);
  if (!item || item.section === 'hospital') return null;
  if (item.perSession && item.perSession > 0) return Math.floor(qty / item.perSession);
  return null;
}

export type StockStatus = 'red' | 'amber' | 'green';

export function stockStatus(code: string, qty: number): StockStatus {
  const item = getItem(code);
  if (!item) return qty > 0 ? 'green' : 'red';

  if (item.section === 'hospital') {
    if (qty <= 0) return 'red';
    if (qty <= 1) return 'amber';
    return 'green';
  }

  const sr = sessionsRemaining(code, qty);
  if (sr === null) return qty <= 0 ? 'red' : qty <= 1 ? 'amber' : 'green';
  if (sr < RED_THRESHOLD) return 'red';
  if (sr < AMBER_THRESHOLD) return 'amber';
  return 'green';
}

export function needsOrdering(code: string, qty: number): boolean {
  const item = getItem(code);
  if (!item || item.section === 'hospital') return false;
  return qty < item.targetQty;
}

export function orderUnits(code: string, currentQty: number): number {
  const item = getItem(code);
  if (!item || item.section === 'hospital') return 0;
  return Math.max(0, item.targetQty - currentQty);
}

export function orderBoxes(code: string, currentQty: number): number {
  const item = getItem(code);
  if (!item) return 0;
  return Math.ceil(orderUnits(code, currentQty) / item.boxSize);
}

export interface StockEntry { code: string; qty: number; }

export function sortStock(entries: StockEntry[]): StockEntry[] {
  const nxstage = entries.filter(e => getItem(e.code)?.section === 'nxstage');
  const hospital = entries.filter(e => getItem(e.code)?.section === 'hospital');
  const unknowns = entries.filter(e => !getItem(e.code));

  function sortGroup(group: StockEntry[]): StockEntry[] {
    return [...group].sort((a, b) => {
      const aNeedsOrder = needsOrdering(a.code, a.qty);
      const bNeedsOrder = needsOrdering(b.code, b.qty);
      if (aNeedsOrder !== bNeedsOrder) return aNeedsOrder ? -1 : 1;
      return (getItem(a.code)?.priority ?? 99) - (getItem(b.code)?.priority ?? 99);
    });
  }

  return [...sortGroup(nxstage), ...sortGroup(hospital), ...unknowns];
}

File: flutter/lib/features/inventory/stock_calc.dart
import 'constants.dart';
import 'rate_overrides.dart';

/// Port of frontend/src/routes/Inventory/lib/stockCalc.ts.

const _redThreshold = 8; // < 2 weeks of sessions
const _amberThreshold = 16; // < 4 weeks of sessions

enum StockStatus { red, amber, green }

int? sessionsRemaining(String code, int qty,
    {Map<String, RateOverride> rates = const {}}) {
  if (code == 'PAK-001') return qty * 10;  // always hardcoded, no override
  // Needles: 2 per session — overridable
  if (code == 'P00012326') {
    final o = rates['P00012326']?.perSession;
    if (o != null && o > 0) return qty ~/ o;
    return qty ~/ 2;
  }
  final item = resolveItem(code, rates);
  if (item == null || item.section == 'hospital') return null;
  if (item.perSession != null && item.perSession! > 0) {
    return qty ~/ item.perSession!;
  }
  return null;
}

StockStatus stockStatus(String code, int qty,
    {Map<String, RateOverride> rates = const {}}) {
  final item = resolveItem(code, rates) ?? getItem(code);
  if (item == null) return qty > 0 ? StockStatus.green : StockStatus.red;

  if (item.section == 'hospital') {
    if (qty <= 0) return StockStatus.red;
    if (qty <= 1) return StockStatus.amber;
    return StockStatus.green;
  }

  final sr = sessionsRemaining(code, qty, rates: rates);
  if (sr == null) {
    return qty <= 0
        ? StockStatus.red
        : qty <= 1
            ? StockStatus.amber
            : StockStatus.green;
  }
  if (sr < _redThreshold) return StockStatus.red;
  if (sr < _amberThreshold) return StockStatus.amber;
  return StockStatus.green;
}

bool needsOrdering(String code, int qty,
    {Map<String, RateOverride> rates = const {}}) {
  final item = resolveItem(code, rates);
  if (item == null || item.section == 'hospital') return false;
  return qty < item.targetQty;
}

int consumedUnits(String code, int sessionsTotal,
    {Map<String, RateOverride> rates = const {}}) {
  if (code == 'PAK-001') return (sessionsTotal / 10).ceil();  // always hardcoded, no override
  // Needles: 2 per session — overridable
  if (code == 'P00012326') {
    final o = rates['P00012326']?.perSession;
    if (o != null && o > 0) return sessionsTotal * o;
    return sessionsTotal * 2;
  }
  final item = resolveItem(code, rates);
  if (item == null || item.perSession == null || item.perSession! <= 0) return 0;
  return sessionsTotal * item.perSession!;
}

int orderUnits(String code, int currentQty,
    {int backupQty = 0,
    int? deliverySessions,
    Map<String, RateOverride> rates = const {}}) {
  final item = resolveItem(code, rates);
  if (item == null || item.section == 'hospital') return 0;
  final working = (currentQty - backupQty).clamp(0, currentQty);
  int target;
  if (deliverySessions != null) {
    final consumed = consumedUnits(code, deliverySessions, rates: rates);
    target = consumed > 0 ? consumed : item.targetQty;
  } else {
    target = item.targetQty;
  }
  final n = target - working;
  return n < 0 ? 0 : n;
}

int orderBoxes(String code, int currentQty,
    {int backupQty = 0,
    int? deliverySessions,
    Map<String, RateOverride> rates = const {}}) {
  final item = resolveItem(code, rates) ?? getItem(code);
  if (item == null) return 0;
  return (orderUnits(code, currentQty,
              backupQty: backupQty,
              deliverySessions: deliverySessions,
              rates: rates) /
          item.boxSize)
      .ceil();
}

class StockEntry {
  const StockEntry(this.code, this.qty);
  final String code;
  final int qty;
}

List<StockEntry> sortStock(List<StockEntry> entries,
    {Map<String, RateOverride> rates = const {}}) {
  List<StockEntry> sortGroup(List<StockEntry> group) {
    final g = [...group];
    g.sort((a, b) {
      final aNeeds = needsOrdering(a.code, a.qty, rates: rates);
      final bNeeds = needsOrdering(b.code, b.qty, rates: rates);
      if (aNeeds != bNeeds) return aNeeds ? -1 : 1;
      return (getItem(a.code)?.priority ?? 99)
          .compareTo(getItem(b.code)?.priority ?? 99);
    });
    return g;
  }

  final nxstage =
      entries.where((e) => getItem(e.code)?.section == 'nxstage').toList();
  final hospital =
      entries.where((e) => getItem(e.code)?.section == 'hospital').toList();
  final unknown = entries.where((e) => getItem(e.code) == null).toList();
  return [...sortGroup(nxstage), ...sortGroup(hospital), ...unknown];
}
